### PR TITLE
Enhance snapshot for blocked entry

### PR DESCRIPTION
**Pull request by luckyxiaoqiang**
### Describe what this PR does / why we need it
Enhance block error.

### Does this pull request fix one issue?
#282 

#### Block error spec
- MUST specify block type.
- MUST specify related rule.
- MUST specify the snapshot value if the `RuleCheckSlot` compares with a specific metric when do checking.
- SUGGEST a simply and important message to describe the block error.

### CODE REVIEW

File: core/flow/tc_default.go [hunk 1]
```diff
@@ -47,7 +49,10 @@ func (d *RejectTrafficShapingChecker) DoCheck(resStat base.StatNode, batchCount
 	}
 	curCount := float64(metricReadonlyStat.GetSum(base.MetricEventPass))
 	if curCount+float64(batchCount) > threshold {
-		return base.NewTokenResultBlockedWithCause(base.BlockTypeFlow, "", d.rule, curCount)
+		interval := d.rule.StatIntervalInMs
+		msg := fmt.Sprintf("rate check not pass, rule id: %s, stat interval: %d ms, current count: %.2f, request batch count: %d, threshold: %.2f",
```
Comment (sanxun0325) [inline core/flow/tc_default.go:53]: Log print parameters. The KeysAndValues method is recommended

File: core/flow/tc_throttling.go [hunk 2]
```diff
@@ -44,8 +45,24 @@ func (c *ThrottlingChecker) DoCheck(_ base.StatNode, batchCount uint32, threshol
 	if batchCount <= 0 {
 		return nil
 	}
+
+	var rule *Rule
+	var ruleID string
+	if c.BoundOwner() != nil {
+		rule = c.BoundOwner().BoundRule()
+	}
+	if rule != nil {
```
[the remark this answers is not in the document]
Comment (luckyxiaoqiang) [reply core/flow/tc_throttling.go:54]: `BoundOwner` would be nil in unit test, so rule would be nil too. Maybe we should refine unit test.

File: core/flow/tc_throttling.go [hunk 3]
```diff
@@ -44,8 +45,24 @@ func (c *ThrottlingChecker) DoCheck(_ base.StatNode, batchCount uint32, threshol
...
+		ruleID = rule.ID
+	}
+
 	if threshold <= 0.0 {
-		return base.NewTokenResultBlocked(base.BlockTypeFlow)
+		msg := fmt.Sprintf("rate check not pass, threshold is <= 0.0, rule id: %s", ruleID)
+		if rule == nil {
+			// Should only appear in unit test
+			return base.NewTokenResultBlockedWithMessage(base.BlockTypeFlow, msg)
```
[the remark this answers is not in the document]
Comment (luckyxiaoqiang) [reply core/flow/tc_throttling.go:62]: `rule` would be nil in unit test.

File: core/flow/tc_throttling.go [hunk 4]
```diff
@@ -63,15 +80,31 @@ func (c *ThrottlingChecker) DoCheck(_ base.StatNode, batchCount uint32, threshol
...
 	if estimatedQueueingDuration > c.maxQueueingTimeNs {
-		return base.NewTokenResultBlocked(base.BlockTypeFlow)
+		msg := fmt.Sprintf("rate check not pass, throttling estimated queueing time exceeds max queueing time, "+
+			"rule id: %s, estimated: %.2f ms, max: %.2f ms",
+			ruleID, float64(estimatedQueueingDuration)/float64(util.UnixTimeUnitOffset), float64(c.maxQueueingTimeNs)/float64(util.UnixTimeUnitOffset))
+		if rule == nil {
+			// Should only appear in unit test
+			return base.NewTokenResultBlockedWithMessage(base.BlockTypeFlow, msg)
+		} else {
+			return base.NewTokenResultBlockedWithCause(base.BlockTypeFlow, msg, rule, nil)
```
[the remark this answers is not in the document]
Comment (luckyxiaoqiang) [reply core/flow/tc_throttling.go:90]: There might be too many similar functions. I'm thinking about using `Options` to create `TokenResult`. Any advice on that?

File: core/flow/tc_throttling.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (sanxun0325) [reply core/flow/tc_throttling.go:54]: Maybe in the NewThrottlingChecker, the owner should add a check that cannot be nil

File: core/circuitbreaker/slot.go [hunk 5]
```diff
@@ -31,10 +31,11 @@ func (b *Slot) Check(ctx *base.EntryContext) *base.TokenResult {
 		return result
 	}
 	if passed, rule := checkPass(ctx); !passed {
+		msg := "circuit breaker check not pass"
```
Comment (louyuting) [inline core/circuitbreaker/slot.go:34]: Maybe "circuit breaker check  blocked" is better?

Comment (luckyxiaoqiang) [reply core/circuitbreaker/slot.go]: Updated.

**Review by louyuting — APPROVED** [review]
Nice work and looking forward more.

**Pull request merged**